### PR TITLE
[TT-7658] implement OAS ID extractor

### PR DESCRIPTION
**Pull request by jeffy-mathew**
https://tyktech.atlassian.net/browse/TT-7658

- Implement OAS ID extractor 
- Strict type ID extractor config in classic API defintion

### CODE REVIEW

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4755/merge
Commit:  
Triggered by: pull_request (@jeffy-mathew)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/4141236326)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4755/merge
Commit: 3abade54777cfb56e82817e3fef8530c533aac73 
Triggered by: pull_request (@jeffy-mathew)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/4141237279)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4755/merge
Commit: c492eed1345aab2d2ab489cfe5615171f3ddc1f9 
Triggered by: pull_request (@jeffy-mathew)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/4141338680)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4755/merge
Commit: 43ab4d46037bb45075581efd1368d733e825f55b 
Triggered by: pull_request (@jeffy-mathew)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/4141433913)

File: apidef/oas/schema/x-tyk-gateway.md [hunk 1]
```diff
@@ -529,6 +529,42 @@ Path is the path to shared object file in case of gopluign mode or path to js co
 **Field: `rawBodyOnly` (`boolean`)**
 RawBodyOnly if set to true, do not fill body in request or response object.
 
+**Field: `idExtractor` ([IDExtractor](#idextractor))**
+IDExtractor  configures ID extractor with coprocess custom authentication.
```
Comment (titpetric) [inline apidef/oas/schema/x-tyk-gateway.md:533]: I'm looking at the double space here and I don't know, is it our generator? doesn't gofmt clean it from comments?

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4755/merge
Commit: 9931ae333efceeb0eac2c5891a5d7db4ce1a97bc 
Triggered by: pull_request (@jeffy-mathew)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/4142201728)

File: apidef/oas/authentication.go [hunk 2]
```diff
@@ -647,18 +647,97 @@ type AuthenticationPlugin struct {
...
+func (id *IDExtractorConfig) Fill(api apidef.APIDefinition) {
+	id.HeaderName = api.CustomMiddleware.IdExtractor.ExtractorConfig.HeaderName
+	id.FormParamName = api.CustomMiddleware.IdExtractor.ExtractorConfig.FormParamName
+	id.Regexp = api.CustomMiddleware.IdExtractor.ExtractorConfig.RegexExpression
+	id.RegexpMatchIndex = api.CustomMiddleware.IdExtractor.ExtractorConfig.RegexMatchIndex
+	id.XPathExp = api.CustomMiddleware.IdExtractor.ExtractorConfig.XPathExpression
+}
+
+func (id *IDExtractorConfig) ExtractTo(api *apidef.APIDefinition) {
+	api.CustomMiddleware.IdExtractor.ExtractorConfig.HeaderName = id.HeaderName
```
Comment (titpetric) [inline apidef/oas/authentication.go:705]: This looks to be nearly the full object definition... why not overwrite it with *val = T{}?

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4755/merge
Commit: 9dd07400a234b722dfdf6e4d9319c5c83996c51c 
Triggered by: pull_request (@jeffy-mathew)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/4142345779)

File: apidef/oas/schema/x-tyk-gateway.md (hunk 1 above)
Comment (jeffy-mathew) [reply apidef/oas/schema/x-tyk-gateway.md:533]: It's because of the comments, updating it.

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4755/merge
Commit: 5173c218f6112202e3d6e4b6ea9a9a36dff8f6c1 
Triggered by: pull_request (@jeffy-mathew)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/4142508320)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4755/merge
Commit: 573baec6c6fd803b3c3d984baeb43dfa5f683a9a 
Triggered by: pull_request (@jeffy-mathew)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/4142569888)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4755/merge
Commit: e0fd1d11cad7b5f31f3d5e5e307f7abd895cbd8c 
Triggered by: pull_request (@jeffy-mathew)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/4142563632)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4755/merge
Commit: e941341f7464dec775295c8fa36986de587198b9 
Triggered by: pull_request (@jeffy-mathew)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/4142650757)

**Review by titpetric — APPROVED** [review]

**Review by Yasintrm — APPROVED** [review]

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4755/merge
Commit: ced95bb0457d48f358063a38aa5f3e5814aef263 
Triggered by: pull_request (@jeffy-mathew)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/4142844133)

**Pull request merged**